packages/core/src/compiler/Style.ts

```diff
@@ -3046,11 +3077,103 @@ export const translate = (
 
 //#endregion
 
+//#region group graph
+
+export const checkGroupGraph = (groupGraph: GroupGraph): StyleWarning[] => {
+  const warnings: StyleWarning[] = [];
+  for (const name of groupGraph.nodes()) {
+    if (groupGraph.parents(name).length > 1) {
+      warnings.push({
+        tag: "ShapeBelongsToMultipleGroups",
+        shape: name,
+        groups: groupGraph.parents(name),
+      });
+    }
+  }
+
+  const cycles = groupGraph.findCycles();
+  if (cycles.length !== 0) {
+    warnings.push({
+      tag: "GroupCycleWarning",
+      cycles,
+    });
+  }
+
+  return warnings;
+};
+
+//#endregion
+
 //#region layering
 
+export const propagateLayeringToParents = (
+  { below, above }: Layer,
+  layerGraph: Graph<string>,
+  groupGraph: GroupGraph
+): void => {
+  layerGraph.setEdge({ i: below, j: above, e: undefined });
+  const belowParents = groupGraph.parents(below);
+  const aboveParents = groupGraph.parents(above);
+  aboveParents
+    .filter((p) => p !== below)
+    .forEach((p) => {
+      propagateLayeringToParents({ below, above: p }, layerGraph, groupGraph);
```

Comment:
A bit unfortunate to do this with recursion, since we pretty much avoid recursion elsewhere throughout the compiler... not the end of the world though, I'm just curious if there's a way to do it without recursion (since we are operating over an explicit graph).